### ISSUE TITLE
Early fusion in MAG-BERT

### ISSUE THREAD

**Issue by ky941122** [WasifurRahman/BERT_multimodal_transformer#9]
Hi, I found that you hard code the MAG layer just after the first embedding layer in MAG_BertModel, what's the reason for it. Thanks.

**Comment by RE-N-Y** [comment]
Sure, one of the reason why we injected the MAG layer right after initial embedding layer is for 2 reasons.

1. In BERT, initial embedding layer provides word embedding + positional encoding. In case of XLNet, similarly, both word embedding + positional information is encoded after its first transformer layer. Hence, it makes sense to inject MAG after those respective layers

2. As we showed in Table 2 of section 6.2 in our paper, we noticed that early fusion tends to perform better than late fusion strategy. My guess is that multi modal features needs to be propagated a few layers. We recently used MAG with Albert in other paper, but it doesn't seem to work as well as MAG-BERT / MAG-XLNet because Albert uses weight-sharing strategy between layers.

<img width="378" alt="Screen Shot 2020-11-09 at 10 03 05 PM" src="https://user-images.githubusercontent.com/19684973/98544437-5be55000-22d7-11eb-9743-bb3f8928a3b5.png">

**Comment by ky941122** [comment]
Thanks for replying. 
So can I understand that adding the MAG layer in this position is optimal for BERT?  The paper does not mention the results corresponding to BERT.

**Comment by RE-N-Y** [comment]
I believe so, we examined injection position for XLNet because initially, we attempted injection before the first transformer layer (i.e. no positional information was encoded). And that didn't give us good results and decided to examine how MAG injection position influences the performance. However, yes, we didn't examine optimal position for BERT, but from MAG-XLNET I doubt trying late fusion would yield significantly better results. But, if you do end up examining it, I would love to see the results!
